external/kafka/src/main/scala/org/apache/spark/streaming/kafka/DirectKafkaInputDStream.scala

```diff
@@ -158,4 +166,37 @@ class DirectKafkaInputDStream[
     }
   }
 
+  private[streaming]
+  class DirectKafkaStreamingListener extends StreamingListener {
+
+    val offsetCommitEnabled = context.conf.getBoolean("spark.streaming.kafka.offsetCommit", false)
+
+    val zkClientOpt = if (offsetCommitEnabled) {
+      kafkaParams.get("zookeeper.connect").map { zkConnect =>
+        val zkSessionTimeoutMs = kafkaParams.getOrElse("zookeeper.session.timeout.ms", "6000").toInt
+        val zkConnectionTimeoutMs =
+          kafkaParams.getOrElse("zookeeper.session.timeout.ms", "6000").toInt
+        new ZkClient(zkConnect, zkSessionTimeoutMs, zkConnectionTimeoutMs, ZKStringSerializer)
+      }
+    } else {
+      None
+    }
+
+    val groupId = "directKafkaConsumer"
+
+    override def onBatchCompleted(batchCompleted: StreamingListenerBatchCompleted) = synchronized {
```

Comment:
Because ZK is not the system of record, are we really gaining much for the added complexity of onBatchCompleted, as opposed to just updating ZK in compute() ?